core/os/midi_driver.cpp

```diff
@@ -42,77 +42,85 @@ void MIDIDriver::set_singleton() {
 	singleton = this;
 }
 
-void MIDIDriver::receive_input_packet(int device_index, uint64_t timestamp, uint8_t *data, uint32_t length) {
-	Ref<InputEventMIDI> event;
-	event.instantiate();
-	event->set_device(device_index);
-	uint32_t param_position = 1;
-
-	if (length >= 1) {
-		if (data[0] >= 0xF0) {
-			// channel does not apply to system common messages
-			event->set_channel(0);
-			event->set_message(MIDIMessage(data[0]));
-			last_received_message = data[0];
-		} else if ((data[0] & 0x80) == 0x00) {
-			// running status
-			event->set_channel(last_received_message & 0xF);
-			event->set_message(MIDIMessage(last_received_message >> 4));
-			param_position = 0;
-		} else {
-			event->set_channel(data[0] & 0xF);
-			event->set_message(MIDIMessage(data[0] >> 4));
-			param_position = 1;
-			last_received_message = data[0];
-		}
-	}
+void MIDIDriver::receive_input_packet(int p_device_index, uint64_t p_timestamp, uint8_t *p_data, uint32_t p_length) {
+	// p_data may contain multiple messages, eg multiple note on using MIDI 'running status'.
+	uint32_t index = 0; // index will be incremented as each status or data byte is read from p_data.
 
-	switch (event->get_message()) {
-		case MIDIMessage::AFTERTOUCH:
-			if (length >= 2 + param_position) {
-				event->set_pitch(data[param_position]);
-				event->set_pressure(data[param_position + 1]);
-			}
-			break;
+	// Each time through loop will consume a MIDI message worth of bytes from p_data and create one InputEventMIDI.
+	while (index < p_length) {
+		Ref<InputEventMIDI> event;
+		event.instantiate();
+		event->set_device(p_device_index);
 
-		case MIDIMessage::CONTROL_CHANGE:
-			if (length >= 2 + param_position) {
-				event->set_controller_number(data[param_position]);
-				event->set_controller_value(data[param_position + 1]);
+		if (p_length - index >= 1) {
+			if (p_data[index] >= 0xF0) {
+				// Channel does not apply to system common messages.
+				event->set_channel(0);
+				event->set_message(MIDIMessage(p_data[index]));
+				last_received_message = p_data[index];
+				index++;
```

Comment:
If this is system exclusive message, will it go out of sync if the data after the first byte is not recognised?

Reply:
I don't know, I have only enough understanding of MIDI spec and of midi_driver.cpp to fix the bug that I have reproduced.  If your question requires investigation, can that be recorded as a separate issue from this PR, as this PR is to fix defect https://github.com/godotengine/godot/issues/77035 ?

Reply:
OK, having read what a system exclusive is https://blog.landr.com/midi-sysex/ I think you are correct. The method should return if it sees one. I don't know what the parsing behavior should be for the packet after the system exclusive. Because this PR ads the while, it should abort. Good catch! :-)

Reply:
Well, if your PR puts this data into a read loop, so if it gets out of sync, anything could happen (it probably aborts later if not recognized, but we need to be sure of this). So if system exclusive is not being responded to it might make sense to abort here.

It's important for PR author to be confident of this (and have the answers, as reviewers might not be familiar), because we don't want to exchange one bug for another bug (or set of bugs).

Reply:
It is literally decades since I touched MIDI streams, so I only have very rough memories.

https://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html

> F0 <length> <bytes to be transmitted after F0>

> The length is stored as a variable-length quantity. It specifies the number of bytes which follow it, not including the F0 or the length itself. For instance, the transmitted message F0 43 12 00 07 F7 would be stored in a MIDI File as F0 05 43 12 00 07 F7. It is required to include the F7 at the end so that the reader of the MIDI File knows that it has read the entire message.

Reply:
I agree. I am going to look into a unit test for MIDI packet parsing, which may mean I will be refactoring MIDIDriver::receive_input_packet into two methods. So if I disappear for a while that's what I'm doing.

Reply:
I started a unit test for midi_driver. When `MIDIDriver::receive_input_packet` is called from a unit test, it can parse midi correctly, but when it comes to `Input::get_singleton()->parse_input_event(event);` it crashes, see below. Should unit tests not be calling deeper into the Godot Engine with `parse_input_event`? Is there some unit test setup that needs to be done to support this? I could refactor the code to separate `parse_input_event` from the midi byte parsing code so the midi parsing can be tested by itself, but this would make my PR a bigger change. @lawnjelly are you able to advise? Thanks. 

`handle_crash: Program crashed with signal 11
Engine version: Godot Engine v4.3.dev.custom_build (46bddb63b8e8f8bbba7aa4383cf4e9e868d1f5a5)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[1] 1   libsystem_platform.dylib            0x000000019f683584 _sigtramp + 56
[2] std::__1::recursive_mutex::lock()
[3] std::__1::recursive_mutex::lock()
[4] std::__1::unique_lock<std::__1::recursive_mutex>::unique_lock[abi:un170006](std::__1::recursive_mutex&) (in godot.macos.editor.dev.arm64) (unique_lock.h:41)
[5] std::__1::unique_lock<std::__1::recursive_mutex>::unique_lock[abi:un170006](std::__1::recursive_mutex&) (in godot.macos.editor.dev.arm64) (unique_lock.h:40)
[6] MutexLock<MutexImpl<std::__1::recursive_mutex>>::MutexLock(MutexImpl<std::__1::recursive_mutex> const&) (in godot.macos.editor.dev.arm64) (mutex.h:81)
[7] MutexLock<MutexImpl<std::__1::recursive_mutex>>::MutexLock(MutexImpl<std::__1::recursive_mutex> const&) (in godot.macos.editor.dev.arm64) (mutex.h:81)
[8] Input::parse_input_event(Ref<InputEvent> const&) (in godot.macos.editor.dev.arm64) (input.cpp:993)
[9] MIDIDriver::receive_input_packet(int, unsigned long long, unsigned char*, unsigned int) (in godot.macos.editor.dev.arm64) (midi_driver.cpp:127)`

Reply:
That sounds like better for a separate PR imo.

Reply:
There is a new problem though. If I flood the midi channel with many pitch change messages memory grows and the process crashes. Same test on Godot4.2 doesn't have this problem.

Reply:
Pitch change problem solved. Parsing error from my change - pitch change is 2 data bytes. (Which original code was doing correctly.)

Reply:
@lawnjelly This MIDI PR is ready for review. Thanks.